### DISCUSSION TITLE
Secret for Authentication Plugin

### DISCUSSION THREAD

**Discussion by mopinon** [verdaccio/verdaccio#2735] [Q&A]
Hi, I want to make an authentication plugin for unity.
I looked at the code of several existing authentication plugins.  They all use some kind of `secret` to create an `authToken`.
All plugins get this `secret` from verdaccio, I understand correctly that this `secret` comes from the `config.yaml` file?

**Comment by edclement** [comment]
Hi @mopinon, with Verdaccio@5, the storage plugin handles setting/getting the secret: https://github.com/verdaccio/monorepo/blob/v10.0.6/plugins/local-storage/src/local-database.ts#L37. The default `local-storage` plugin simply reads/writes it from the `.verdaccio-db.json` file from a property named `secret`.

I have not investigated how this is handled in Verdaccio@6

**Comment by juanpicado** [comment]
> I understand correctly that this secret comes from the config.yaml file?

Follow up of @edclement , it is generated on startup once the `.verdaccio-db.json` is created, either on v5 or v6, @edclement frankly I'd like to make it more flexible, but not sure how would be the best practice here (security) since the token is used to sign tokens.

**Reply by edclement** [reply]
> frankly I'd like to make it more flexible

@juanpicado I personally think having the retrieval/persistence of the token delegated to the storage plugin makes it flexible enough. One thing I have noticed however is that on startup (at least in Verdaccio@5):
1. Verdaccio retrieves the JWT secret value.
2. If:
    a. If the value exists, it uses that value
    b. If the value does NOT exist, it generates a new secret value
3. Verdaccio then **always** invokes the storage plugins `setSecret` function.

If the secret value already exists, the call to `setSecret` really should NOT occur. This isn't a big deal with the default `local-storage` plugin. However, for implementations of the storage plugin that use managed cloud services (e.g. GCP Secret Manager or AWS Secret Manager) this can become a problem. In those scenarios you wouldn't want to give the compute resources Verdaccio is running on write access to the secret (hopefully for obvious reasons).

**Reply by juanpicado** [reply]
🤔 yeah, that does not sound good, I think now is a good moment to revert that behavior and make it right.
